src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.plugin.query;
+
+import java.io.IOException;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.commons.lang.builder.EqualsBuilder;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.apache.lucene.search.Query;
+import org.opensearch.common.ParseField;
+import org.opensearch.common.ParsingException;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.xcontent.XContentBuilder;
+import org.opensearch.common.xcontent.XContentParser;
+import org.opensearch.index.mapper.NumberFieldMapper;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.QueryShardContext;
+
+/**
+ * NeuralQueryBuilder is responsible for producing "neural" query types. A "neural" query type is a wrapper around a
+ * k-NN vector query. It uses a ML language model to produce a dense vector from a query string that is then used as
+ * the query vector for the k-NN search.
+ */
+
+@Log4j2
+@Getter
+@NoArgsConstructor
+public class NeuralQueryBuilder extends AbstractQueryBuilder<NeuralQueryBuilder> {
+
+    public static final String NAME = "neural";
+
+    static final ParseField QUERY_TEXT_FIELD = new ParseField("query_text");
+
+    static final ParseField MODEL_ID_FIELD = new ParseField("model_id");
+
+    static final ParseField K_FIELD = new ParseField("k");
+
+    private String fieldName;
+    private String queryText;
+    private String modelId;
+    private int k;
+
+    /**
+     * Set the fieldName this query will be executed against
+     *
+     * @param fieldName name of k-NN vector field that query will be executed against
+     * @return this
+     */
+    public NeuralQueryBuilder fieldName(String fieldName) {
+        this.fieldName = fieldName;
+        return this;
+    }
```

Comment:
I thought about this, but I didnt see how it would work, given we are extending a builder class. Instead, I followed how AbstractQueryBuilder handled it for the boost field: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractQueryBuilder.java#L172